File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-8.py
import os

class Solution:
    def findMin(self, nums: List[int]) -> int:
        l, r = 0, len(nums) - 1
        while l <= r:
            m = l + (r - l) // 2
            if right_clean := nums[m] < nums[r]:
                r = m
            else:
                l = m + 1

        return nums[m]


# Scenario (0)
# [1, 2, [3], 4, 5]
# [[1], 2]
# [1]

# Scenario (1)
# [5, 6, [1], 2, 3, 4]
# [5, [6], 1]
# [[6], 1]


# Scenario (2)
# [3, 4, [5], 6, 1, 2]
# [5, [6], 1, 2]
# [6, [1], 2]
# 

# Scenario (3)
# [6, 1, [2], 3, 4, 5]
# [6, [1], 2]
# [[6], 1]
# [1]

# Scenario (4)
# [3, 4, [5], 1, 2]
# 

File: Data Structures & Algorithms/top-k-elements-in-list/submission-8.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> m;
        for (auto& num : nums) {
            m[num]++;
        }

        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        for (auto& [num, freq] : m) {
            pq.push(pair<int, int>{freq, num});
            if (pq.size() > k) {
                pq.pop();
            }
        }

        vector<int> result;
        while (!pq.empty()) {
            auto [freq, num] = pq.top();
            result.push_back(num);
            pq.pop();
        }

        return result;
    }
};
